lib/OpenQA/WebAPI/Controller/Step.pm

```diff
@@ -91,6 +94,45 @@ sub view ($self) {
     $self->viewimg;
 }
 
+sub _checkout_needledir_or_default {
+    my $job = shift;
+    my $tmp_needles_dir = $job->needle_dir;
+    my $real_needle_dir = realpath($tmp_needles_dir) // $tmp_needles_dir;
+    if (my $casedir = $job->settings->single({key => 'CASEDIR'})) {
+        my $casedir_refspec = Mojo::URL->new($casedir->value)->fragment;
+        eval {
+            my $vars_json = Mojo::File->new($job->result_dir(), 'vars.json')->slurp;
+            my $vars = decode_json($vars_json);
+            $casedir_refspec = $vars->{TEST_GIT_HASH};
```

Comment:
The needles might be in a separate repository, e.g. this is the case for the openSUSE test distribution and thus our main production use case. This means the `TEST_GIT_HASH` and the `CASEDIR`-fragment are not relevant. The `NEEDLES_GIT_REFSPEC` and `NEEDLES_DIR`-fragment would be the relevant variables which should be checked before. Note that there is also `TEST_GIT_REFSPEC` which could be checked as well.

Maybe it would make sense to introduce `NEEDLES_GIT_HASH` so we could also do the same lookup for needles.